### DISCUSSION TITLE
Please add type-2 polar generation

### DISCUSSION THREAD

**Discussion by kjayawar** [peterdsharpe/AeroSandbox#114] [Ideas]
Hi Peter,

I've been using ASB and NeuralFoil for last couple of weeks and LOVED it. 
However, I found the alpha based aerodynamic coefficient really limiting at times. 

I tried 2 approaches so far to remedy this -- but none of these are ideal

1. I tried linearized aerodynamic calculations to figure out the Alpha corresponding to CL0 and then got remaining alpha from Cl_alpha of 2pi -> then got the CL to scale Re and calculated polar.
2. Other method is a simple 2 step process where I calculated aero using a given alpha range and then use the incoming CL to scale RE
aero_t1 = af.get_aero_from_neuralfoil(alpha=alpha_range, Re=Re)
aero_t2 = af.get_aero_from_neuralfoil(alpha=alpha_range, Re=Re/aero_t1['CL'])

My question is, is it possible to modify get_aero_from_kulfan_parameters(..) such that it takes list of CLs and Re numbers instead of list of Alphas and Re numbers ? I have a gut feeling it should be possible to do this by changing the rows and columns of the NN 

Thanks very much for your consideration :)

As a side note: May I also ask which software you used to train this NN ? 

Best
Kanishka.

**Comment by peterdsharpe** [comment]
Hi Kanishka,

I'm glad to hear you're having a great time using AeroSandbox/NeuralFoil!

-----

On specified-$C_L$ airfoil analysis:

Using NeuralFoil, there's no easy way to transform the existing neural network to do this as an explicit calculation. There are also many non-neural-network calculations that NeuralFoil does (seen in the source of `asb.KulfanAirfoil.get_aero_from_neuralfoil()`) which you'd also have to invert, so this would be a pretty tricky task.

Doing specified-$C_L$ analysis is also a problematic task in the first place, especially if the goal is to optimize through this calculation - this is because $C_L$ does not uniquely map back to a single angle of attack.

The most obvious example of this is the periodicity of $\alpha$ (i.e., $\alpha = 5 \degree$ is the same as $\alpha = 365 \degree$), but other less-trivial examples exist: for example, $d(C_L)/d\alpha$ sometimes goes negative after stall. But it's not limited to this - $d(C_L)/d\alpha$ can go negative way before stall, and it's not obvious what airfoil that will happen with.

For example, consider this airfoil below ([`*.dat` file attached here; you'll have to change the file extension since GitHub only allows `*.txt`](https://github.com/peterdsharpe/AeroSandbox/files/13812935/pathological.dat.txt))

![image](https://github.com/peterdsharpe/AeroSandbox/assets/6295985/594f11ab-40c6-4794-aaea-a399a728141b)

Looks pretty normal! But it has a pretty wild $C_{L\alpha}$ polar, shown below at $\text{Re} = 33123$.

![image](https://github.com/peterdsharpe/AeroSandbox/assets/6295985/a9229a1a-ccb6-42ef-b6d5-caa4bd461572)

So, if XFoil (or NeuralFoil, since it's trained on XFoil) were asked to analyze this airfoil at a fixed $C_L=0$, which of the three intersections would you get? In reality, whichever one corresponded to the initial guess's [basis of Newton attraction](https://acme.byu.edu/0000017a-1bb8-db63-a97e-7bfa0bec0000/newtons-pdf) - basically, it's very hard to predict. This is fine for human-in-the-loop analysis, but you can see how an optimization process would freak out with a multi-valued solution like this.

Specifying alpha is, in general, better-behaved. Multi-valued alpha solutions do exist in nature (as an example, stall hysteresis), but are much rarer than multi-valued CL solutions.

-----

With that said, the best way to specify $C_L$ for the purposes of analysis are:

1. If you're doing pure analysis, wrap the `asb.KulfanAirfoil.get_aero_from_neuralfoil()` call with something like [`scipy.optimize.fsolve()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fsolve.html), and iteratively solve for the $C_L$ of your choosing.
2. If you're doing AeroSandbox optimization, set $\alpha$ as an optimization variable, and apply a constraint on the resulting $C_L$.

If you have an example of what you're trying to do with a specified-$C_L$ polar, there may be other suggestions or more specific advice that can help further.
